### ISSUE TITLE
Metadata update handling

### ISSUE THREAD

**Issue by phirestalker** [bonukai/MediaTracker#634]
Right now some shows are not updated if the updates include watched seasons. I realize that you are working on this for the 1.0 update. The short description of the fix sounds like it may still leave out some updates by only updating the new unwatched seasons.

I have a few questions about the data available from TMDB and the handling of show metadata update (i.e. season and episode number changes)

1. Does TMDB have episode IDs as well as show IDs?
2. If not, do they at least have absolute episode numbers?
3. If they have either of those available, wouldn't it be possible to update the metadata and keep the watched status?

I guess if they do not have either of those things, I will need to make some suggestions. The amount of tomfoolery over which episodes are in what season over there is out of control (especially for some shows, notorious in fact).

**Comment by bonukai** [comment]
TMDB have episodes and seasons IDs, and in the new version they are used to detect change of episode numbers. Just like you suggest, if a season cannot be updated due to watched episodes (this usually happens, when two episodes are merged into one), then only this season is not updated. The show metadata, and other season are updated in this situation.

I am still open to suggestions :)

**Comment by bonukai** [comment]
> I was just thinking that since the episodes have some way to uniquely identify them, other than their season and episode, seasons that are watched could still be updated or rearranged. Or, do you mean that the only time it can't be updated is when they merge episodes?

Season won't be updated if there is an episode in local database marked as seen or rated, but this episode has been removed from the TMDB. In this case only affected season is not updated.

If episodes have changed number, then they will be rearranged, even if they have been marked as seen, based on the episode TMDB id.

> Also, I want to clarify. You mean shows update unwatched seasons in the wip branch, but still not for the main branch right?

Yes, this is only implemented in the `wip` branch.

> This brings up another question. How is it tracked when I watch an episode in Jellyfin? Is it by the episode id? When they change seasons (like after DVD releases) can subsequent watches get linked to the wrong episode?

Jellyfin plugin only uses season and episode number to identify an episode in the MediaTracker

I have never came across situation, where season number would be rearranged, so this was not implemented. Do you have an example of a show, when this happen?

**Comment by phirestalker** [comment]
I have an example of a show where it is likely to have or to happen. Futurama. They have changed the seasons from being 12 seasons to being only 9, so there must be some shuffling between seasons.

**Comment by phirestalker** [comment]
When this fix is complete, will I need to do anything special, or will it just start updating shows that I have watched previous seasons of?

**Comment by bonukai** [comment]
Update will be seamless, everything will update itself.

**Comment by phirestalker** [comment]
Since there is no (at least per last discussion) way forward for episodes that need to be updated in watched seasons, would you consider adding an export function? I am thinking I can export the show (or the whole shebang if necessary) delete all watches, and then update all the shows. Once that is done I can then re-import the watch status backup.

Would this work like I think it would? I realize it might miss some episodes on import if they were changed.

I need to study the API and schema for TMDB. I would hope that since they do have an episode ID, that they keep some sort of a history of changes. I just wonder if they make it public. If they do, then that could be checked to match watched status for episodes they have changed.

**Comment by bradnak** [comment]
I know this is a massive bump, but TMDB *do* have an API for tracking database changes  - see https://developer.themoviedb.org/docs/tracking-content-changes and the links on that page. By default, it returns changes in the last 24 hours, but it is configurable.

**Comment by phirestalker** [comment]
Use ryot. I switched and it is great.
